### PR TITLE
Request features

### LINKED ISSUE #293: Request Features
ASP.NET 5->Fundamentals->Request Features
- Feature collection and relationship to HttpContext
- Feature collection mutability: added, removed, wrapped and replaced
- Created by servers **and** middleware
    - Servers own creating the collection
    - Middleware adds to the collection using `HttpContext.SetFeature<T>(feature)`
- Consumed by middleware and app code (see `HttpContext.GetFeature<T>()`)
- Built-in features (https://github.com/aspnet/HttpAbstractions/tree/dev/src/Microsoft.AspNet.Http.Features)
    - Note: feature specifics should be covered in API ref docs (coming soon)

### PR DESCRIPTION
**Pull request by ardalis**
Fixes #293

### CODE REVIEW

**Comment by danroth27** [comment]
@Tratcher Please review

File: aspnet/fundamentals/request-features.rst [hunk 1]
```diff
@@ -1,8 +1,177 @@
-.. include:: /../common/stub-topic.txt
+Request Features
+================
 
-|stub-icon| Request Features
-============================
+By `Steve Smith`_
 
-.. include:: /../common/stub-notice.txt
+Individual web server features related to how HTTP requests and responses are handled have been factored into separate interfaces, defined in the `HttpAbstractions <https://github.com/aspnet/HttpAbstractions>`_ package. These abstractions are used by individual server implementations and middleware to create and modify the application's hosting pipeline.
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: HttpAbstractions is a github repository, not a nuget package. The package is Microsoft.AspNet.Http.Features.

File: aspnet/fundamentals/request-features.rst [hunk 2]
```diff
@@ -1,8 +1,177 @@
...
-.. _issue: https://github.com/aspnet/Docs/issues/293
+In this article:
+	- `Feature interfaces`_
+	- `Feature collections`_
+	- `Middleware and request features`_
+	
+Feature interfaces
+------------------
+
+ASP.NET 5 defines a number of `Http Feature Interfaces <https://github.com/aspnet/HttpAbstractions/tree/dev/src/Microsoft.AspNet.Http.Features>`_, which are used by servers to identify which features they support. The most basic features of a web server are the ability to handle requests and return responses, as defined by the following feature interfaces:
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: Link to lables, not the dev branch:
https://github.com/aspnet/HttpAbstractions/tree/1.0.0-beta5/src/Microsoft.AspNet.Http.Features

File: aspnet/fundamentals/request-features.rst [hunk 3]
```diff
@@ -1,8 +1,177 @@
...
+
+`IHttpRequestFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IHttpRequestFeature.cs>`_
+	Defines the structure of an HTTP request, including the protocol, path, QueryString, headers, and body.
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: nit: the casing of QueryString is inconsistent.

File: aspnet/fundamentals/request-features.rst [hunk 4]
```diff
@@ -1,8 +1,177 @@
...
+	Defines support for `HTTP Upgrades <http://tools.ietf.org/html/rfc2616#section-14.42>`_, which allow the client to specify which additional protocols it would like to use if the server wishes to switch protocols.
+
+`IHttpBufferingFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IHttpBufferingFeature.cs>`_
+	Defines methods for disabling buffering of requests and/or responses.
+
+`IHttpConnectionFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IHttpConnectionFeature.cs>`_
+	Defines properties for local and remote addresses and ports.
+
+`IHttpRequestLifetimeFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IHttpRequestLifetimeFeature.cs>`_
+	Defines support for aborting connections.
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: Or detecting if a request has been terminated prematurely, such as by a client disconnect.

File: aspnet/fundamentals/request-features.rst [hunk 5]
```diff
@@ -1,8 +1,177 @@
...
+	Defines a method for sending files asynchronously.
+
+`IHttpWebSocketFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IHttpWebSocketFeature.cs>`_
+	Defines an API for supporting web sockets.
+
+`IRequestIdentifierFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/IRequestIdentifierFeature.cs>`_
+	Adds a property that can be implemented to uniquely identify requests.
+
+`ISessionFeature <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Features/ISessionFeature.cs>`_
+	Defines ``ISessionFactory`` and ``ISession`` abstractions for supporting user sessions.
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: This is out of date.

File: aspnet/fundamentals/request-features.rst [hunk 6]
```diff
@@ -1,8 +1,177 @@
...
+	
+Feature collections
+-------------------
+
+The HttpAbstractions repository includes a FeatureModel package. Its main ingredient is the `FeatureCollection <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.FeatureModel/FeatureCollection.cs>`_ type, which is used frequently by :doc:`servers` and their requests, as well as :doc:`middleware`, to identify which features they support. The `HttpContext <https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Abstractions/HttpContext.cs>`_ type defined in ``Microsoft.AspNet.Http.Abstractions`` (not to be confused with the ``HttpContext`` defined in ``System.Web``) provides an interface for getting and setting these features. The `OwinEnvironment <https://github.com/aspnet/HttpAbstractions/blob/5fe8037281bb826e0708abdcdafbc76571dc21f5/src/Microsoft.AspNet.Owin/OwinEnvironment.cs>`_ class provides some examples of how application code can verify the presence of needed features, and react accordingly.
```
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: Don't use OwinEnvironment as an example, most people don't know what it's trying to do. Use this example instead:
https://github.com/aspnet/StaticFiles/blob/1.0.0-beta5/src/Microsoft.AspNet.StaticFiles/StaticFileContext.cs#L327-L331

Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: Hmm, you already use the StaticFile example below. I think you can just remove this section about OwinEnvironment.

**Comment by Tratcher** [comment]
:shipit: after addressing minor comments.

File: aspnet/fundamentals/request-features.rst (hunk 1 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Fixed.

File: aspnet/fundamentals/request-features.rst (hunk 2 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Fixed.

File: aspnet/fundamentals/request-features.rst (hunk 3 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: with what? there is only one instance of QueryString in the document. Do you prefer it Querystring?

File: aspnet/fundamentals/request-features.rst (hunk 4 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Updated.

File: aspnet/fundamentals/request-features.rst (hunk 5 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Updated to link to beta5 label; interface looks the same as described. What is out of date?

File: aspnet/fundamentals/request-features.rst (hunk 3 above)
Comment (Tratcher) [inline aspnet/fundamentals/request-features.rst]: with all the other parameters in this sentence. `query string` might fit better.

Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Ah, ok, fixed.

File: aspnet/fundamentals/request-features.rst (hunk 6 above)
Comment (ardalis) [inline aspnet/fundamentals/request-features.rst]: Removed;condensed section without reference to OwinEnvironment.

**Comment by ardalis** [comment]
@Tratcher ship it or does ISessionFeature need something still?

**Comment by Tratcher** [comment]
Nevermind, the ISessionFeature refactor was in beta6.

**Pull request merged**